### DISCUSSION TITLE
How to inject Quarkus managed ExecutorService and ScheduledExecutorService?

### DISCUSSION THREAD

**Discussion by nimo23** [quarkusio/quarkus#39096] [Q&A] [answered]
In Wildfly I can inject the Executor services managed by Wildfly as follows:

```
// cast from ManagedExecutorService to ExecutorService
@Resource(lookup = "java:jboss/ee/concurrency/executor/default")
private ExecutorService executor;

// cast from ManagedScheduledExecutorService to ScheduledExecutorService
@Resource(lookup = "java:jboss/ee/concurrency/scheduler/default")
private ScheduledExecutorService scheduler;
```

How can I access/inject the Quarkus managed `ExecutorService` and `ScheduledExecutorService`?

[a comment by an automated account is not shown]

**Comment by manovotn** [comment]
You can inject a managed executor:

```
import org.eclipse.microprofile.context.ManagedExecutor;

// ...

@Inject
ManagedExecutor exec;
```

The above is a basic version. You can modify it too in terms of queue, max async tasks and what contexta should it propagate.

As for scheduled tasks, I'd suggest looking at quarkus-scheduler which might be what you ultimately want to achieve?
See https://quarkus.io/guides/scheduler-reference

**Reply by nimo23** [reply]
Would this also be possible?:

```
@Inject // from Arc directly
private ExecutorService executor;

@Inject // from Arc directly
private ScheduledExecutorService scheduler;
```

I need these 2 instances. Since I only have `ManagedExecutor` (as you suggested), I have to cast it to `ExecutorService` for my class, and I'm still missing a quarkus managed `ScheduledExecutorService`. 

According to https://quarkus.io/guides/virtual-threads, an quarkus managed executor service can also be injected as follows:

```
@Inject
@VirtualThreads
ExecutorService vThreads;
```

So I assumed that we can do it also that way:

```
@Inject // from Arc directly
private ExecutorService executor;

@Inject // from Arc directly
private ScheduledExecutorService scheduler;
```

But I'm not sure if that's okay.

**Reply by nimo23** [reply]
I tried it with

```
@Inject // from Arc directly
private ExecutorService executor;

@Inject // from Arc directly
private ScheduledExecutorService scheduler;
```

and it seems it works. I logged the injected class instances:


```
2024-03-01 01:27:26,989 INFO  [io.app] (Quarkus Main Thread) executor = class io.quarkus.arc.generator.ScheduledExecutorService_cAsQhMqDYUnGxT4AN4njjudQ50I_Synthetic_ClientProxy
2024-03-01 01:27:26,989 INFO  [io.app] (Quarkus Main Thread) scheduler = class io.quarkus.arc.generator.ScheduledExecutorService_cAsQhMqDYUnGxT4AN4njjudQ50I_Synthetic_ClientProxy
```

As you can see, the instances for `executor` and `scheduler` are the same. Does this mean that Quarkus uses the same instance for both `ManagedExecutor` and `ScheduledExecutorService`? It's a bit irritating because in wildfly the `executor` and the `scheduler` are two different instances.

**Reply by mkouba** [reply, marked as answer]
> As you can see, the instances for executor and scheduler are the same. Does this mean that Quarkus uses the same instance for both ManagedExecutor and ScheduledExecutorService? It's a bit irritating because in wildfly the executor and the scheduler are two different instances.

@nimo23 Yes, you can `@Inject ExecutorService` and `@Inject ScheduledExecutorService` and it's expected that in both cases you get the same instance (or rather a client proxy delegating to the same instance), i.e. the default executor for blocking tasks.

And indeed, the `@Inject @VirtualThreads ExecutorService` injects an executor to run tasks on virtual threads.

The `@Inject ManagedExecutor` mentioned by Matej is a specialized executor that allows you to specify various aspects of context propagation (as defined in the  MicroProfile Context Propagation).

**Reply by mschorsch** [reply]
@nimo23 see also https://quarkus.io/guides/context-propagation#usage-example-for-completionstage

**Reply by nimo23** [reply]
>The @Inject ManagedExecutor mentioned by Matej is a specialized executor that allows you to specify various aspects of context propagation (as defined in the MicroProfile Context Propagation).

Using `@Inject ManagedExecutor` does not return the (managed) `ScheduledExecutorService` but the `ExecutorService`:

`managedExecutor = class io.smallrye.context.SmallRyeManagedExecutor_N80A159oF8msb0BBZG1sW4wdTGk_Synthetic_ClientProxy`

So I must use
```

@Inject // from Arc directly
private ExecutorService executor;

@Inject // from Arc directly
private ScheduledExecutorService scheduler;
```

instead of 

`@Inject ManagedExecutor`.
